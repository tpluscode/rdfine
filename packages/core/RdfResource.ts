/* eslint-disable camelcase,@typescript-eslint/camelcase,no-dupe-class-members */
import RDF from '@rdfjs/data-model'
import { NamespaceBuilder } from '@rdfjs/namespace'
import { NamedNode, DatasetCore, BlankNode, DefaultGraph, Quad_Graph, Term, Literal } from 'rdf-js'
import cf, { SafeClownface, SingleContextClownface } from 'clownface'
import ResourceFactoryImpl, {
  Constructor,
  Mixin,
  ResourceCreationOptions,
  ResourceFactory,
  ResourceIndexer,
} from './lib/ResourceFactory'
import once from 'once'
import TypeCollectionCtor, { TypeCollection } from './lib/TypeCollection'
import { xsd } from '@tpluscode/rdf-ns-builders'

type ObjectOrFactory<T> = T | ((self: RdfResource) => T)

export type ResourceIdentifier = BlankNode | NamedNode
export type ResourceNode<D extends DatasetCore = DatasetCore> = { dataset: D; term: Term; graph?: NamedNode | DefaultGraph }

export interface GetOptions {
  strict: boolean
}

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: ResourceIdentifier
  readonly types: TypeCollection<D>
  readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  readonly _graphId: Quad_Graph
  readonly _parent?: RdfResource<D>
  readonly isAnonymous: boolean
  hasType (type: string | NamedNode): boolean
  /**
   * Gets the value of a property
   * @param property
   */
  get<T extends RdfResource = RdfResource> (property: string | NamedNode): T
  get<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T | null
  /**
   * Gets the value of a property and ensures that an array will be returned
   * @param property
   */
  getArray<T extends RdfResource = RdfResource> (property: string | NamedNode, options?: GetOptions): T[]
  /**
   * Gets the property value if it's boolean. Throws if it's not
   * @param property
   */
  getBoolean (property: string | NamedNode, options?: GetOptions): boolean
  /**
   * Gets the property value if it's a string. Throws if it's not
   * @param property
   */
  getString (property: string | NamedNode): string
  getString (property: string | NamedNode, options?: GetOptions): string | null
  /**
   * Gets the property value if it's a number. Throws if it's not
   * @param property
   */
  getNumber (property: string | NamedNode): number | null
  getNumber (property: string | NamedNode, options?: GetOptions): number | null
  _getObjects(property: string | NamedNode, options?: GetOptions): SafeClownface<Term, D>
  _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options?: ResourceCreationOptions<D>): T & ResourceIndexer
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _selfGraph: SingleContextClownface<ResourceIdentifier, D>
  public readonly _unionGraph: SafeClownface<ResourceIdentifier, D>
  public readonly __initialized: boolean = false
  public readonly _parent?: RdfResource<D>
  private readonly __initializeProperties: (() => boolean)
  public static __ns?: NamespaceBuilder
  public static factory: ResourceFactory = new ResourceFactoryImpl(RdfResourceImpl)
  public static __mixins: Mixin[] = []

  private static _userInitializeProperties(resource: RdfResourceImpl, init: Initializer<RdfResource> = {}): void {
    Object.entries(init)
      .filter(([prop]) => prop !== 'id' && prop !== 'types')
      .forEach(([prop, value]) => {
        (resource as any)[prop] = value
      })

    if (init.types && Array.isArray(init.types)) {
      init.types.forEach(type => resource.types.add(type))
    }
  }

  public constructor(graph: ResourceNode<D>, init: Initializer<any> = {}, parent?: RdfResource<D>) {
    if (graph.term.termType !== 'BlankNode' && graph.term.termType !== 'NamedNode') {
      throw new Error(`RdfResource cannot be initialized from a ${graph.term.termType} node`)
    }

    const selfGraph = cf({
      ...graph,
      term: graph.term,
    })

    if (selfGraph._context[0].graph) {
      this._selfGraph = selfGraph
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: undefined })
    } else {
      this._selfGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term, graph: RDF.defaultGraph() })
      this._unionGraph = cf({ dataset: selfGraph.dataset, term: selfGraph.term })
    }

    this.__initializeProperties = once(() => {
      const self = this as any
      const defaults: Map<string, ObjectOrFactory<any>> = self.constructor.__defaults || new Map<string, ObjectOrFactory<any>>()
      defaults.forEach((value, key) => {
        const currentValue = self[key]
        const valueIsEmptyArray = Array.isArray(currentValue) && currentValue.length === 0
        const valueIsUndefined = typeof currentValue === 'undefined'

        if (!(valueIsEmptyArray || valueIsUndefined)) {
          return
        }

        if (typeof value === 'function') {
          self[key] = value(self)
          return
        }

        self[key] = value
      })

      return true
    })

    this._parent = parent
    this.__initialized = this.__initializeProperties()
    RdfResourceImpl._userInitializeProperties(this, init)
  }

  public get id(): ResourceIdentifier {
    return this._selfGraph.term
  }

  public get _graphId(): Quad_Graph {
    return this._selfGraph._context[0].graph!
  }

  public get types(): TypeCollection<D> {
    return new TypeCollectionCtor(this)
  }

  public get isAnonymous() {
    return this.id.termType === 'BlankNode'
  }

  public hasType(type: string | NamedNode): boolean {
    return this.types.has(type)
  }

  public equals(other: RdfResource | undefined | null): boolean {
    if (!other) {
      return false
    }

    return this.id.equals(other.id)
  }

  public get<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T | RdfResource<D> | null {
    const objects = this.getArray<T>(property, options)

    if (objects.length > 0) {
      return objects[0]
    }

    return null
  }

  public getArray<T extends RdfResource = RdfResource>(property: string | NamedNode, options?: GetOptions): T[] {
    const values = this._getObjects(property, options)
      .filter(obj => obj.term.termType === 'NamedNode' || obj.term.termType === 'BlankNode')
      .map(obj => {
        return this._create<T>(obj, [], { parent: this })
      })

    if (!values.length) {
      return []
    }

    return values
  }

  public getNumber(property: string | NamedNode): number
  public getNumber(property: string | NamedNode, options?: GetOptions): number | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return parseFloat(value.value)
    }

    throw new Error(`Expected property '${property}' to be a number but found '${value}'`)
  }

  public getString(property: string | NamedNode): string
  public getString(property: string | NamedNode, options?: GetOptions): string | null {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return null
    }

    if (value.term.termType === 'Literal') {
      return value.value
    }

    throw new Error(`Expected property '${property}' to be a literal but found '${value}'`)
  }

  public getBoolean(property: string | NamedNode, options?: GetOptions): boolean {
    const [value] = this._getObjects(property, options).toArray()

    if (typeof value === 'undefined') {
      return false
    }

    if (value.term.termType === 'Literal' && xsd.boolean.equals(value.term.datatype)) {
      return value.term.equals(this._selfGraph.literal(true).term)
    }

    throw new Error(`Expected property '${property}' to be a boolean but found '${value}'`)
  }

  public _getObjects(property: string | NamedNode, { strict }: GetOptions = { strict: false }): SafeClownface<Term, D> {
    const propertyNode = typeof property === 'string' ? this._selfGraph.namedNode(property) : property
    const objects = this._selfGraph.out(propertyNode)

    if (!objects.terms.length && strict) {
      throw new Error(`Value for predicate <${property}> was missing`)
    }

    return objects
  }

  public _create<T extends RdfResource>(term: ResourceNode<D>, mixins?: Mixin[] | [Constructor, ...Mixin[]], options: ResourceCreationOptions<D> = {}): T & ResourceIndexer {
    return (this.constructor as Constructor).factory.createEntity<T>(term, mixins, options)
  }
}

type UserDefinedInterface<T extends RdfResource> = Omit<T, keyof RdfResource>

interface BaseInitializer {
  types?: NamedNode[] | TypeCollection<any>
  id?: RdfResource['id']
}

type InitialNode<Node extends Term = NamedNode | BlankNode> = Node | SingleContextClownface<Node>
type InitialLiteral = InitialNode<Literal>

export type Initializer<T> = Partial<Omit<{
  [P in keyof T]?:
  T[P] extends (infer U)[] ? U extends RdfResource ? Initializer<UserDefinedInterface<U>>[] | InitialNode[]
    : U extends Term ? T[P] | InitialNode<Term>[] : T[P] | InitialLiteral[]
    : T[P] extends RdfResource ? Initializer<UserDefinedInterface<T[P]> & BaseInitializer> | InitialNode :
      T[P] extends Term ? T[P] | InitialNode<T[P]>
        : T[P] | InitialLiteral
}, keyof RdfResource>> & BaseInitializer

type PartialRecursive<T> = T extends object ? { [K in keyof T]?: PartialRecursive<T[K]> } : T

export function fromObject<T extends RdfResource>(initializer: Initializer<UserDefinedInterface<T> & BaseInitializer>): T {
  return initializer as unknown as T
}
