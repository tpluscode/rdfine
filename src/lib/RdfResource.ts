import { Term, NamedNode, DatasetCore, BlankNode } from 'rdf-js'
import cf, { SingleContextClownface } from 'clownface'
import ns from '@rdfjs/namespace'
import once from 'once'

const rdf = ns('http://www.w3.org/1999/02/22-rdf-syntax-ns#')

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: Term | null
  readonly types: Term[]
  readonly _node: SingleContextClownface<D>
  hasType (type: string | NamedNode): boolean
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _node: SingleContextClownface<D>
  private readonly __initializeProperties: (() => void)
  public static __ns?: any
  private static __defaults = new Map<string, unknown>()

  public constructor(node: SingleContextClownface<D> | { dataset: D; term: NamedNode | BlankNode; graph?: NamedNode }) {
    const contexts = cf(node).toArray()

    if (contexts.length !== 1) {
      throw new Error('RdfResource can only be initialized from a single node. Got ' + contexts.length)
    }

    this._node = cf(node).toArray()[0]

    this.__initializeProperties = once(() => {
      const self = this as any
      RdfResourceImpl.__defaults.forEach((value, key) => {
        if (typeof self[key] === 'undefined') {
          self[key] = value
        }
      })
    })

    this.__initializeProperties()
  }

  public get id() {
    if (!this._node.term) throw new Error('Graph context does not represent a single node')

    return this._node.term
  }

  public get types() {
    return this._node.out(rdf.type).terms
  }

  public hasType(type: string | NamedNode) {
    const typeNode = typeof type === 'string' ? this._node.namedNode(type) : type

    return this._node.has(rdf.type, typeNode).terms.length > 0
  }
}
