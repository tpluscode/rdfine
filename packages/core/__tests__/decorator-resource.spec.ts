import { prefixes } from '@zazuko/rdf-vocabularies'
import cf, { SingleContextClownface } from 'clownface'
import { DatasetCore, DefaultGraph, Literal, NamedNode, Term } from 'rdf-js'
import rdfExt from 'rdf-ext'
import DatasetExt from 'rdf-ext/lib/Dataset'
import { defaultGraph, literal } from '@rdfjs/data-model'
import { turtle } from '@tpluscode/rdf-string'
import {
  property,
  namespace,
  crossBoundaries,
  Constructor,
  RdfResource as IRdfResource,
} from '../index'
import RdfResource, { fromObject } from '../RdfResource'
import ResourceFactoryImpl from '../lib/ResourceFactory'
import { parse, vocabs } from './_helpers'

const { ex, foaf, schema, rdf } = vocabs

describe('decorator', () => {
  describe('resource', () => {
    describe('getter', () => {
      it('returns a resource object using path predicate', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:friend ex:friend .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: foaf.friend })
          friend!: RdfResource
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.friend

        // then
        expect(friend).toBeInstanceOf(RdfResource)
        expect(friend.id.value).toEqual(ex.friend.value)
      })

      it('can be used with namespace decorator', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:friend ex:friend .
        `)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property.resource()
          friend!: RdfResource
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.friend

        // then
        expect(friend).toBeInstanceOf(RdfResource)
        expect(friend.id.value).toEqual(ex.friend.value)
      })

      it('returns rdf list array', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:knows ( ex:Will ex:Joe ex:Sindy ) .
        `)
        class Resource extends RdfResource {
          @property.resource({ path: foaf.knows, values: 'list' })
          friends!: RdfResource[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.friends.map(l => l.id.value)).toEqual([
          ex.Will.value,
          ex.Joe.value,
          ex.Sindy.value,
        ])
      })

      it('sets itself as parent of created resource', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:friend ex:friend .
        `)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property.resource()
          friend!: RdfResource
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.friend

        // then
        expect(friend._parent).toBe(instance)
      })
    })

    describe('setter', () => {
      it('sets a relation between resources', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:john a schema:Person .
          ex:jane a schema:Person .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: schema.spouse })
          spouse?: RdfResource
        }
        const john = new Resource({ dataset, term: ex.john })
        const jane = new Resource({ dataset, term: ex.jane })

        // when
        john.spouse = jane

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('setting null removes relation', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:john a schema:Person .
          ex:jane a schema:Person .
          
          ex:john schema:spouse ex:jane .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: schema.spouse })
          spouse?: RdfResource | null
        }
        const john = new Resource({ dataset, term: ex.john })

        // when
        john.spouse = null

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('accepts raw named node term', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:john a schema:Person .
          ex:jane a schema:Person .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: schema.spouse })
          spouse?: RdfResource | NamedNode
        }
        const john = new Resource({ dataset, term: ex.john })

        // when
        john.spouse = ex.jane

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('accepts blank node context', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:john a schema:Person .
          ex:jane a schema:Person .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: schema.spouse })
          spouse?: RdfResource | SingleContextClownface
        }
        const john = new Resource({ dataset, term: ex.john })

        // when
        john.spouse = john._selfGraph.blankNode()

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets array to rdf list property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:john foaf:knows (
            ex:stacy
            ex:frank
          ) .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: foaf.knows, values: 'list' })
          knows?: (RdfResource | NamedNode)[]
        }
        const john = new Resource({ dataset, term: ex.john })

        // when
        john.knows = [
          ex.jane,
          ex.joe,
        ]

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets array to set property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:john foaf:knows ex:stacy, ex:frank .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: foaf.knows, values: 'array' })
          knows?: (RdfResource | NamedNode)[]
        }
        const john = new Resource({ dataset, term: ex.john })

        // when
        john.knows = [
          ex.jane,
          ex.joe,
        ]

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('accepts raw blank node term', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:lois a schema:Person .
          _:clark a ex:Superman .
        `)

        class Resource extends RdfResource {
          @property.resource({ path: schema.spouse })
          spouse?: RdfResource | Term
        }
        const lois = new Resource({ dataset, term: ex.lois })

        // when
        lois.spouse = cf({ dataset }).has(rdf.type, ex.Superman).term

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('accepts deep plain JS object', async () => {
        // given
        interface Person extends IRdfResource {
          knows: Person
          name: string
          parents: Person[]
        }
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:Lois a schema:Person .
        `)
        function Mixin<B extends Constructor>(base: B) {
          class Resource extends base implements Person {
            @property.resource({ path: schema.knows })
            knows!: Person

            @property.literal({ path: schema('name') })
            name!: string

            @property.resource({ path: schema.parent, values: 'array' })
            parents!: Person[]
          }
          return Resource
        }
        Mixin.shouldApply = () => true
        class TestResourceBase extends RdfResource {}
        TestResourceBase.factory = new ResourceFactoryImpl(TestResourceBase)
        TestResourceBase.factory.addMixin(Mixin)
        const lois = TestResourceBase.factory.createEntity<Person>({ dataset, term: ex.Lois })

        // when
        lois.knows = fromObject({
          id: ex.Superman,
          types: [schema.Person],
          name: 'Superman',
          parents: [{
            name: 'Jor-El',
          }, {
            name: 'Lara Lor-Van',
          }],
        })

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })
    })

    describe('initial', () => {
      it('set object returned by factory function', async () => {
        // given
        const dataset = rdfExt.dataset()
        @namespace('http://example.com/')
        class NameResource extends RdfResource {
          @property.literal()
          first!: string

          @property.literal()
          last!: string

          @property.resource()
          person!: RdfResource
        }
        class Resource extends RdfResource {
          @property.resource({
            path: foaf.friend,
            initial: (self: Resource) => {
              const name = new NameResource(self._selfGraph.blankNode())
              name.first = 'John'
              name.last = 'Doe'
              name.person = self
              return name
            },
            as: [NameResource],
          })
          name!: NameResource
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const name = instance.name

        // then
        expect(name).toBeInstanceOf(RdfResource)
        expect(name.first).toEqual('John')
        expect(name.last).toEqual('Doe')
        expect(name.person.id.value).toEqual(instance.id.value)
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('set by named node', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.resource({
            path: schema.employee,
            initial: ex.Google,
          })
          employer!: RdfResource
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.employer.id).toEqual(ex.Google)
      })

      it('set object returned by factory function as clownface', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.resource({
            path: foaf.friend,
            initial: (self: Resource) => {
              return self._selfGraph.namedNode('http://example.com/friend')
            },
          })
          friend!: RdfResource
        }

        // when
        // eslint-disable-next-line no-new
        new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })
    })

    describe('over multiple graphs', () => {
      class Resource<D extends DatasetCore> extends RdfResource<D> {
        @property.resource({ path: foaf.knows })
        friend!: this

        @property({ path: schema('name') })
        name?: Literal

        @property.resource({
          path: crossBoundaries(foaf.knows),
          as: [Resource],
        })
        allAboutFriends!: this[]

        @property.resource({
          path: crossBoundaries(foaf.knows),
          subjectFromAllGraphs: true,
          as: [Resource],
        })
        allAboutAllFriends!: this[]

        @property.resource({
          path: foaf.knows,
          subjectFromAllGraphs: true,
          as: [Resource],
        })
        allFriends!: this[]
      }

      function namedGraphTests(newResource: (dataset: DatasetExt, term: NamedNode, graph?: NamedNode | DefaultGraph) => Resource<DatasetExt>) {
        describe('getter', () => {
          it('does not cross named graph boundary by default', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              @prefix schema: <${prefixes.schema}> .
              
              ex:John foaf:knows ex:Will ex:John .
              
              ex:Will schema:name "William" ex:Will .
            `)

            // when
            const instance = newResource(dataset, ex.John, ex.John)

            // then
            expect(instance.friend.name).toBeUndefined()
          })

          it('crosses named graph boundary when explicitly annotated', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              @prefix schema: <${prefixes.schema}> .
              
              ex:John foaf:knows ex:Will ex:John .
              
              ex:Will schema:name "William" ex:Will .
              ex:Will foaf:knows ex:John ex:Will .
            `)

            // when
            const instance = newResource(dataset, ex.John, ex.John)

            // then
            expect(instance.allAboutFriends).toHaveLength(2)
            const friends = instance.allAboutFriends
            expect(friends.map(will => will._graphId)).toEqual(
              expect.arrayContaining([ex.John, ex.Will]),
            )
            expect(friends.map(will => will.name)).toEqual(
              expect.arrayContaining([literal('William')]),
            )
          })

          it('when crossing named graph boundaries returns separate resource object for each named graph', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              @prefix schema: <${prefixes.schema}> .
              
              ex:John foaf:knows ex:Will .
              
              ex:Will schema:name "William" ex:WillName .
              ex:Will foaf:knows ex:John ex:WillFriends .
              ex:Will schema:employee ex:Google ex:WillJob .
            `)

            // when
            const instance = newResource(dataset, ex.John)

            // then
            expect(instance.allAboutFriends).toHaveLength(4)
            const wills = instance.allAboutFriends
            expect(wills.map(w => w._graphId)).toEqual(
              expect.arrayContaining([defaultGraph(), ex.WillName, ex.WillFriends, ex.WillJob]),
            )
          })

          it('when crossing named graph boundaries does not merge subject from other graphs', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              @prefix schema: <${prefixes.schema}> .
              
              ex:John foaf:knows ex:Will .
              ex:John foaf:knows ex:Sindy ex:JohnGraph .              
            `)

            // when
            const instance = newResource(dataset, ex.John)

            // then
            expect(instance.name).toBeUndefined()
            expect(instance.allAboutFriends).toHaveLength(1)
          })

          it('when crossing named graph boundaries can also merge subject from other graphs', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              @prefix schema: <${prefixes.schema}> .
              
              ex:John foaf:knows ex:Will .
              ex:John schema:name "John Doe" ex:JohnGraph .
              ex:John foaf:knows ex:Sindy ex:JohnGraph .    
            `)

            // when
            const instance = newResource(dataset, ex.John)

            // then
            expect(instance.allAboutAllFriends).toHaveLength(2)
          })

          it('when merging subjects, preserves the graph', async () => {
            // given
            const dataset = await parse(turtle`
              ${ex.John} a ${schema.Person} .
            
              ${ex.WillGraph} {
                ${ex.John} ${foaf.knows} ${ex.Will} .
                ${ex.Will} ${schema.name} "Will" .
              }
              
              ${ex.SindyGraph} {
                ${ex.John} ${foaf.knows} ${ex.Sindy} .
              }
              
              ${ex.UnreachableGraph} {
                ${ex.Sindy} ${schema.name} "Sindy" .
              }
            `)

            // when
            const instance = newResource(dataset, ex.John)
            const friendNames = instance.allFriends.map(friend => friend.name)

            // then
            expect(friendNames).toEqual(
              expect.arrayContaining([literal('Will'), undefined]),
            )
          })
        })

        describe('setter', () => {
          it('assert link quad in subject\'s graph', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              @prefix schema: <${prefixes.schema}> .
              
              ex:John schema:name "John" ex:John .
              
              ex:Will schema:name "William" ex:Will .
            `)

            // when
            const john = newResource(dataset, ex.John, ex.John)
            john.friend = newResource(dataset, ex.Will, ex.Will)

            // then
            expect(dataset.toCanonical()).toMatchSnapshot()
          })
        })
      }

      describe('constructed from plain object', () => {
        namedGraphTests((dataset, term, graph) => {
          return new Resource({ dataset, term, graph })
        })
      })

      describe('constructed from clownface graph', () => {
        namedGraphTests((dataset, term, graph) => {
          return new Resource(cf({ dataset, term, graph }))
        })
      })
    })
  })
})
