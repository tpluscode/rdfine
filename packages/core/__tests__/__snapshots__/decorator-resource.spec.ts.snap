// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorator resource initial set object returned by factory function 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/friend> _:c14n0 .
_:c14n0 <http://example.com/first> \\"John\\" .
_:c14n0 <http://example.com/last> \\"Doe\\" .
_:c14n0 <http://example.com/person> <http://example.com/res> .
"
`;

exports[`decorator resource initial set object returned by factory function as clownface 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/friend> <http://example.com/friend> .
"
`;

exports[`decorator resource over multiple graphs setter assert link quad in subject's graph 1`] = `
"<http://example.com/John> <http://schema.org/name> \\"John\\" <http://example.com/John> .
<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Will> <http://example.com/John> .
<http://example.com/Will> <http://schema.org/name> \\"William\\" <http://example.com/Will> .
"
`;

exports[`decorator resource setter accepts blank node context 1`] = `
"<http://example.com/jane> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
<http://example.com/john> <http://schema.org/spouse> _:c14n0 .
<http://example.com/john> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
"
`;

exports[`decorator resource setter accepts deep plain JS object 1`] = `
"<http://example.com/Lois> <http://schema.org/knows> <http://example.com/Superman> .
<http://example.com/Lois> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
<http://example.com/Superman> <http://schema.org/name> \\"Superman\\" .
<http://example.com/Superman> <http://schema.org/parent> _:c14n0 .
<http://example.com/Superman> <http://schema.org/parent> _:c14n1 .
<http://example.com/Superman> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
_:c14n0 <http://schema.org/name> \\"Lara Lor-Van\\" .
_:c14n1 <http://schema.org/name> \\"Jor-El\\" .
"
`;

exports[`decorator resource setter accepts raw blank node term 1`] = `
"<http://example.com/lois> <http://schema.org/spouse> _:c14n0 .
<http://example.com/lois> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Superman> .
"
`;

exports[`decorator resource setter accepts raw named node term 1`] = `
"<http://example.com/jane> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
<http://example.com/john> <http://schema.org/spouse> <http://example.com/jane> .
<http://example.com/john> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
"
`;

exports[`decorator resource setter sets a relation between resources 1`] = `
"<http://example.com/jane> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
<http://example.com/john> <http://schema.org/spouse> <http://example.com/jane> .
<http://example.com/john> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
"
`;

exports[`decorator resource setter sets array to rdf list property 1`] = `
"<http://example.com/john> <http://xmlns.com/foaf/0.1/knows> _:c14n1 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <http://example.com/joe> .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <http://example.com/jane> .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:c14n0 .
"
`;

exports[`decorator resource setter sets array to set property 1`] = `
"<http://example.com/john> <http://xmlns.com/foaf/0.1/knows> <http://example.com/jane> .
<http://example.com/john> <http://xmlns.com/foaf/0.1/knows> <http://example.com/joe> .
"
`;

exports[`decorator resource setter setting null removes relation 1`] = `
"<http://example.com/jane> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
<http://example.com/john> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Person> .
"
`;
